In [38]:
import requests
import pandas as pd
import numpy as np
import json

pd.options.display.max_columns = None
pd.options.display.max_rows = None

In [2]:
url = 'https://fantasy.premierleague.com/api/bootstrap-static/'

In [3]:
r = requests.get(url)

In [4]:
dix = r.json()

In [5]:
dix.keys()

dict_keys(['events', 'game_settings', 'phases', 'teams', 'total_players', 'elements', 'element_stats', 'element_types'])

# Keys Details:

### 'elements' : all the players with following details - 
      "assists": 0,
      "bonus": 0,
      "bps": 0,
      "chance_of_playing_next_round": 0,
      "chance_of_playing_this_round": 0,
      "clean_sheets": 0,
      "code": 232351,
      "cost_change_event": 0,
      "cost_change_event_fall": 0,
      "cost_change_start": -1,
      "cost_change_start_fall": 1,
      "creativity": "0.0",
      "creativity_rank": 499,
      "creativity_rank_type": 190,
      "dreamteam_count": 0,
      "element_type": 2,
      "ep_next": "0.0",
      "ep_this": "0.0",
      "event_points": 0,
      "first_name": "Ryan",
      "form": "0.0",
      "goals_conceded": 0,
      "goals_scored": 0,
      "ict_index": "0.0",
      "ict_index_rank": 520,
      "ict_index_rank_type": 189,
      "id": 607,
      "in_dreamteam": false,
      "influence": "0.0",
      "influence_rank": 516,
      "influence_rank_type": 189,
      "minutes": 0,
      "news": "Joined Coventry City on loan until the end of the season on 31/1. - Expected back 01 Aug",
      "news_added": "2020-01-31T20:30:19.344750Z",
      "now_cost": 39,
      "own_goals": 0,
      "penalties_missed": 0,
      "penalties_saved": 0,
      "photo": "232351.jpg",
      "points_per_game": "0.0",
      "red_cards": 0,
      "saves": 0,
      "second_name": "Giles",
      "selected_by_percent": "0.2",
      "special": false,
      "squad_number": null,
      "status": "n",
      "team": 20,
      "team_code": 39,
      "threat": "0.0",
      "threat_rank": 468,
      "threat_rank_type": 181,
      "total_points": 0,
      "transfers_in": 20970,
      "transfers_in_event": 1437,
      "transfers_out": 10919,
      "transfers_out_event": 739,
      "value_form": "0.0",
      "value_season": "0.0",
      "web_name": "Giles",
      "yellow_cards": 0
      
### "element_stats": 
    {
      "label": "Minutes played",
      "name": "minutes"
    },
    {
      "label": "Goals scored",
      "name": "goals_scored"
    },
    {
      "label": "Assists",
      "name": "assists"
    },
    {
      "label": "Clean sheets",
      "name": "clean_sheets"
    },
    {
      "label": "Goals conceded",
      "name": "goals_conceded"
    },
    {
      "label": "Own goals",
      "name": "own_goals"
    },
    {
      "label": "Penalties saved",
      "name": "penalties_saved"
    },
    {
      "label": "Penalties missed",
      "name": "penalties_missed"
    },
    {
      "label": "Yellow cards",
      "name": "yellow_cards"
    },
    {
      "label": "Red cards",
      "name": "red_cards"
    },
    {
      "label": "Saves",
      "name": "saves"
    },
    {
      "label": "Bonus",
      "name": "bonus"
    },
    {
      "label": "Bonus Points System",
      "name": "bps"
    },
    {
      "label": "Influence",
      "name": "influence"
    },
    {
      "label": "Creativity",
      "name": "creativity"
    },
    {
      "label": "Threat",
      "name": "threat"
    },
    {
      "label": "ICT Index",
      "name": "ict_index"
    }

### "element_types":
    {
      "element_count": 73,
      "id": 1,
      "plural_name": "Goalkeepers",
      "plural_name_short": "GKP",
      "singular_name": "Goalkeeper",
      "singular_name_short": "GKP",
      "squad_max_play": 1,
      "squad_min_play": 1,
      "squad_select": 2,
      "sub_positions_locked": [
        12
      ],
      "ui_shirt_specific": true
    },
    {
      "element_count": 224,
      "id": 2,
      "plural_name": "Defenders",
      "plural_name_short": "DEF",
      "singular_name": "Defender",
      "singular_name_short": "DEF",
      "squad_max_play": 5,
      "squad_min_play": 3,
      "squad_select": 5,
      "sub_positions_locked": [],
      "ui_shirt_specific": false
    },
    {
      "element_count": 276,
      "id": 3,
      "plural_name": "Midfielders",
      "plural_name_short": "MID",
      "singular_name": "Midfielder",
      "singular_name_short": "MID",
      "squad_max_play": 5,
      "squad_min_play": 2,
      "squad_select": 5,
      "sub_positions_locked": [],
      "ui_shirt_specific": false
    },
    {
      "element_count": 92,
      "id": 4,
      "plural_name": "Forwards",
      "plural_name_short": "FWD",
      "singular_name": "Forward",
      "singular_name_short": "FWD",
      "squad_max_play": 3,
      "squad_min_play": 1,
      "squad_select": 3,
      "sub_positions_locked": [],
      "ui_shirt_specific": false
    }

### 'events' : gameweek specific overall stats
    "events": [
    {
      "average_entry_score": 65,
      "chip_plays": [
        {
          "chip_name": "bboost",
          "num_played": 128770
        },
        {
          "chip_name": "3xc",
          "num_played": 271367
        }
      ],
      "data_checked": false,
      "deadline_time": "2019-08-09T18:00:00Z",
      "deadline_time_epoch": 1565373600,
      "deadline_time_game_offset": 0,
      "finished": true,
      "highest_score": 142,
      "highest_scoring_entry": 3493085,
      "id": 1,
      "is_current": false,
      "is_next": false,
      "is_previous": false,
      "most_captained": 191,
      "most_selected": 183,
      "most_transferred_in": 1,
      "most_vice_captained": 189,
      "name": "Gameweek 1",
      "top_element": 214,
      "top_element_info": {
        "id": 214,
        "points": 20
      },
      "transfers_made": 0
    }
    
### 'teams': team-specific stats:
    "teams": [
    {
      "code": 3,
      "draw": 0,
      "form": null,
      "id": 1,
      "loss": 0,
      "name": "Arsenal",
      "played": 0,
      "points": 0,
      "position": 0,
      "pulse_id": 1,
      "short_name": "ARS",
      "strength": 4,
      "strength_attack_away": 1170,
      "strength_attack_home": 1170,
      "strength_defence_away": 1200,
      "strength_defence_home": 1150,
      "strength_overall_away": 1240,
      "strength_overall_home": 1180,
      "team_division": null,
      "unavailable": false,
      "win": 0
    }

In [6]:
dix.keys()

dict_keys(['events', 'game_settings', 'phases', 'teams', 'total_players', 'elements', 'element_stats', 'element_types'])

In [7]:
elements_df = pd.DataFrame(dix['elements'])
element_types_df = pd.DataFrame(dix['element_types'])
element_stats_df = pd.DataFrame(dix['element_stats'])
teams_df = pd.DataFrame(dix['teams'])
events_df = pd.DataFrame(dix['events'])

### Element_stats table isnt needed

In [8]:
element_stats_df.head()

label            name
0  Minutes played         minutes
1    Goals scored    goals_scored
2         Assists         assists
3    Clean sheets    clean_sheets
4  Goals conceded  goals_conceded

## Element_types table:
Might be needed to map playing positions to players in elements table.

In [9]:
element_types_df.head()

id  plural_name plural_name_short singular_name singular_name_short  \
0   1  Goalkeepers               GKP    Goalkeeper                 GKP   
1   2    Defenders               DEF      Defender                 DEF   
2   3  Midfielders               MID    Midfielder                 MID   
3   4     Forwards               FWD       Forward                 FWD   

   squad_select  squad_min_play  squad_max_play  ui_shirt_specific  \
0             2               1               1               True   
1             5               3               5              False   
2             5               2               5              False   
3             3               1               3              False   

  sub_positions_locked  element_count  
0                 [12]             73  
1                   []            224  
2                   []            276  
3                   []             92

### Cleaning:

In [10]:
element_types_df = element_types_df[['id', 'singular_name_short']]
element_types_df.set_index('id',inplace=True)
element_types_df

singular_name_short
id                    
1                  GKP
2                  DEF
3                  MID
4                  FWD

## Teams table:
Might be needed to map team names to players

In [11]:
teams_df.head()

code  draw  form  id  loss         name  played  points  position  \
0     3     0  None   1     0      Arsenal       0       0         0   
1     7     0  None   2     0  Aston Villa       0       0         0   
2    91     0  None   3     0  Bournemouth       0       0         0   
3    36     0  None   4     0     Brighton       0       0         0   
4    90     0  None   5     0      Burnley       0       0         0   

  short_name  strength team_division  unavailable  win  strength_overall_home  \
0        ARS         4          None        False    0                   1180   
1        AVL         2          None        False    0                   1020   
2        BOU         2          None        False    0                   1020   
3        BHA         2          None        False    0                   1050   
4        BUR         3          None        False    0                   1110   

   strength_overall_away  strength_attack_home  strength_attack_away  \
0                   1240                  1170                  1170   
1                   1050                   970                   980   
2                   1020                   990                  1030   
3                   1010                  1100                  1100   
4                   1180                  1130                  1070   

   strength_defence_home  strength_defence_away  pulse_id  
0                   1150                   1200         1  
1                   1000                   1040         2  
2                   1000                   1050       127  
3                   1040                   1030       131  
4                    970                   1110        43

### Cleaning:

In [12]:
teams_df = teams_df[['id', 'name', 'strength', 'strength_overall_home', \
                     'strength_overall_away', 'strength_attack_home' ,\
                     'strength_attack_away', 'strength_defence_home', 'strength_defence_away']]

teams_df.set_index('id',inplace=True)
teams_df

name  strength  strength_overall_home  strength_overall_away  \
id                                                                           
1          Arsenal         4                   1180                   1240   
2      Aston Villa         2                   1020                   1050   
3      Bournemouth         2                   1020                   1020   
4         Brighton         2                   1050                   1010   
5          Burnley         3                   1110                   1180   
6          Chelsea         4                   1220                   1230   
7   Crystal Palace         3                   1130                   1120   
8          Everton         3                   1100                   1240   
9        Leicester         3                   1210                   1160   
10       Liverpool         5                   1340                   1350   
11        Man City         5                   1330                   1340   
12         Man Utd         4                   1220                   1300   
13       Newcastle         3                   1090                   1100   
14         Norwich         2                    980                   1020   
15   Sheffield Utd         3                   1180                   1200   
16     Southampton         3                   1150                   1040   
17           Spurs         4                   1180                   1240   
18         Watford         3                   1030                   1140   
19        West Ham         2                   1040                   1030   
20          Wolves         4                   1230                   1190   

    strength_attack_home  strength_attack_away  strength_defence_home  \
id                                                                      
1                   1170                  1170                   1150   
2                    970                   980                   1000   
3                    990                  1030                   1000   
4                   1100                  1100                   1040   
5                   1130                  1070                    970   
6                   1240                  1240                   1220   
7                   1080                  1180                   1040   
8                   1100                  1210                   1170   
9                   1150                  1080                   1210   
10                  1310                  1320                   1310   
11                  1260                  1270                   1320   
12                  1260                  1330                   1180   
13                  1030                  1040                    980   
14                  1030                  1040                    990   
15                  1170                  1180                   1050   
16                  1080                  1050                   1110   
17                  1100                  1130                   1200   
18                  1120                  1120                   1020   
19                  1000                  1030                   1030   
20                  1180                  1230                   1200   

    strength_defence_away  
id                         
1                    1200  
2                    1040  
3                    1050  
4                    1030  
5                    1110  
6                    1250  
7                    1040  
8                    1180  
9                    1100  
10                   1330  
11                   1320  
12                   1260  
13                    990  
14                    990  
15                   1110  
16                   1080  
17                   1140  
18                   1150  
19                   1150  
20                   1200

## Events table:
Has gameweek specific overall stats. Might need it.

In [13]:
events_df

id          name         deadline_time  average_entry_score  finished  \
0    1    Gameweek 1  2019-08-09T18:00:00Z                   65      True   
1    2    Gameweek 2  2019-08-17T10:30:00Z                   41      True   
2    3    Gameweek 3  2019-08-23T18:00:00Z                   44      True   
3    4    Gameweek 4  2019-08-31T10:30:00Z                   57      True   
4    5    Gameweek 5  2019-09-14T10:30:00Z                   52      True   
5    6    Gameweek 6  2019-09-20T18:00:00Z                   52      True   
6    7    Gameweek 7  2019-09-28T10:30:00Z                   51      True   
7    8    Gameweek 8  2019-10-05T10:30:00Z                   36      True   
8    9    Gameweek 9  2019-10-19T10:30:00Z                   37      True   
9   10   Gameweek 10  2019-10-25T18:00:00Z                   49      True   
10  11   Gameweek 11  2019-11-02T11:30:00Z                   53      True   
11  12   Gameweek 12  2019-11-08T19:00:00Z                   48      True   
12  13   Gameweek 13  2019-11-23T11:30:00Z                   49      True   
13  14   Gameweek 14  2019-11-30T11:30:00Z                   51      True   
14  15   Gameweek 15  2019-12-03T18:30:00Z                   49      True   
15  16   Gameweek 16  2019-12-07T11:30:00Z                   54      True   
16  17   Gameweek 17  2019-12-14T11:30:00Z                   56      True   
17  18   Gameweek 18  2019-12-21T11:30:00Z                   38      True   
18  19   Gameweek 19  2019-12-26T11:30:00Z                   49      True   
19  20   Gameweek 20  2019-12-28T11:30:00Z                   54      True   
20  21   Gameweek 21  2020-01-01T11:30:00Z                   48      True   
21  22   Gameweek 22  2020-01-10T19:00:00Z                   57      True   
22  23   Gameweek 23  2020-01-18T11:30:00Z                   44      True   
23  24   Gameweek 24  2020-01-21T18:30:00Z                   53      True   
24  25   Gameweek 25  2020-02-01T11:30:00Z                   47      True   
25  26   Gameweek 26  2020-02-08T11:30:00Z                   56      True   
26  27   Gameweek 27  2020-02-22T11:30:00Z                   45      True   
27  28   Gameweek 28  2020-02-28T19:00:00Z                   26      True   
28  29   Gameweek 29  2020-03-07T11:30:00Z                   51      True   
29  30   Gameweek 30  2020-03-14T11:30:00Z                    0      True   
30  31   Gameweek 31  2020-03-20T19:00:00Z                    0      True   
31  32   Gameweek 32  2020-04-04T10:30:00Z                    0      True   
32  33   Gameweek 33  2020-04-10T18:00:00Z                    0      True   
33  34   Gameweek 34  2020-04-18T10:30:00Z                    0      True   
34  35   Gameweek 35  2020-04-25T10:30:00Z                    0      True   
35  36   Gameweek 36  2020-05-02T13:00:00Z                    0      True   
36  37   Gameweek 37  2020-05-09T13:00:00Z                    0      True   
37  38   Gameweek 38  2020-05-17T13:00:00Z                    0      True   
38  39  Gameweek 30+  2020-06-17T16:00:00Z                   66      True   
39  40  Gameweek 31+  2020-06-23T16:00:00Z                   62      True   
40  41  Gameweek 32+  2020-06-27T10:30:00Z                   50      True   
41  42  Gameweek 33+  2020-07-04T10:30:00Z                   50      True   
42  43  Gameweek 34+  2020-07-07T16:00:00Z                   59      True   
43  44  Gameweek 35+  2020-07-11T10:30:00Z                   42      True   
44  45  Gameweek 36+  2020-07-14T18:15:00Z                   40      True   
45  46  Gameweek 37+  2020-07-18T15:30:00Z                   53      True   
46  47  Gameweek 38+  2020-07-26T14:00:00Z                    0     False   

    data_checked  highest_scoring_entry  deadline_time_epoch  \
0          False              3493085.0           1565373600   
1          False              6106693.0           1566037800   
2          False              5087809.0           1566583200   
3          False              6372519.0           1

In [14]:
events_df = events_df[['id','name','finished','average_entry_score','highest_score',\
                       'chip_plays','most_selected','most_transferred_in',\
                       'top_element','top_element_info','transfers_made',\
                       'most_captained','most_vice_captained']]
events_df.set_index('id',inplace=True)
events_df

name  finished  average_entry_score  highest_score  \
id                                                               
1     Gameweek 1      True                   65          142.0   
2     Gameweek 2      True                   41          119.0   
3     Gameweek 3      True                   44          107.0   
4     Gameweek 4      True                   57          136.0   
5     Gameweek 5      True                   52          128.0   
6     Gameweek 6      True                   52          141.0   
7     Gameweek 7      True                   51          118.0   
8     Gameweek 8      True                   36          100.0   
9     Gameweek 9      True                   37           98.0   
10   Gameweek 10      True                   49          138.0   
11   Gameweek 11      True                   53          134.0   
12   Gameweek 12      True                   48          119.0   
13   Gameweek 13      True                   49          124.0   
14   Gameweek 14      True                   51          124.0   
15   Gameweek 15      True                   49          117.0   
16   Gameweek 16      True                   54          127.0   
17   Gameweek 17      True                   56          131.0   
18   Gameweek 18      True                   38          100.0   
19   Gameweek 19      True                   49          124.0   
20   Gameweek 20      True                   54          108.0   
21   Gameweek 21      True                   48          112.0   
22   Gameweek 22      True                   57          139.0   
23   Gameweek 23      True                   44          119.0   
24   Gameweek 24      True                   53          137.0   
25   Gameweek 25      True                   47          119.0   
26   Gameweek 26      True                   56          127.0   
27   Gameweek 27      True                   45          126.0   
28   Gameweek 28      True                   26          101.0   
29   Gameweek 29      True                   51          130.0   
30   Gameweek 30      True                    0            0.0   
31   Gameweek 31      True                    0            0.0   
32   Gameweek 32      True                    0            0.0   
33   Gameweek 33      True                    0            0.0   
34   Gameweek 34      True                    0            0.0   
35   Gameweek 35      True                    0            0.0   
36   Gameweek 36      True                    0            0.0   
37   Gameweek 37      True                    0            0.0   
38   Gameweek 38      True                    0            0.0   
39  Gameweek 30+      True                   66          165.0   
40  Gameweek 31+      True                   62          143.0   
41  Gameweek 32+      True                   50          141.0   
42  Gameweek 33+      True                   50          133.0   
43  Gameweek 34+      True                   59          145.0   
44  Gameweek 35+      True                   42          151.0   
45  Gameweek 36+      True                   40          108.0   
46  Gameweek 37+      True                   53          139.0   
47  Gameweek 38+     False                    0            NaN   

                                           chip_plays  most_selected  \
id                                                                     
1   [{'chip_name': 'bboost', 'num_played': 128770}...          183.0   
2   [{'chip_name': 'bboost', 'num_played': 114585}...          183.0   
3   [{'chip_name': 'bboost', 'num_played': 79958},...          183.0   
4   [{'chip_name': 'bboost', 'num_played': 72741},...          183.0   
5   [{'chip_name': 'bboost', 'num_played': 58835},...          183.0   
6   [{'chip_name': 'bboost', 'num_played': 47670},...          183.0   
7   [{'chip_name': 'bboost', 'num_played': 65678},...          183.0   
8   [{'chip_name': 'bboost', 'num_played': 54626},...          183.0   
9   [{'chip_name': 'bboost', 'num_played': 57241},...    

### Elements table:
The main table we would be working on: it has data about each player plaing in the league. We would need to extract required columns, map data from other tables(maybe), clean and analyze data.

In [15]:
elements_df.head()

chance_of_playing_next_round  chance_of_playing_this_round    code  \
0                           0.0                         100.0   69140   
1                         100.0                         100.0   98745   
2                         100.0                         100.0  111457   
3                         100.0                         100.0  154043   
4                         100.0                         100.0   39476   

   cost_change_event  cost_change_event_fall  cost_change_start  \
0                  0                       0                 -4   
1                  0                       0                 -3   
2                  0                       0                 -3   
3                  0                       0                 -5   
4                  0                       0                 -2   

   cost_change_start_fall  dreamteam_count  element_type ep_next ep_this  \
0                       4                0             2     0.0     2.7   
1                       3                0             2     2.3     2.8   
2                       3                0             2     3.3     3.8   
3                       5                1             2     0.8     1.3   
4                       2                1             2     0.5     1.0   

   event_points first_name form  id  in_dreamteam  \
0             0   Shkodran  1.7   1         False   
1             0     Héctor  1.8   2         False   
2             1       Sead  2.8   3         False   
3             0    Ainsley  0.3   4         False   
4             0   Sokratis  0.0   5         False   

                                     news                   news_added  \
0  Hamstring injury - Unknown return date  2020-07-18T22:30:20.553834Z   
1                                          2019-12-09T20:00:21.228098Z   
2                                          2020-02-23T18:30:13.672943Z   
3                                          2019-09-22T18:00:10.824841Z   
4                                          2020-06-18T18:00:15.974146Z   

   now_cost       photo points_per_game       second_name selected_by_percent  \
0        51   69140.jpg             2.9           Mustafi                 0.4   
1        52   98745.jpg             2.9          Bellerín                 1.0   
2        52  111457.jpg             2.2         Kolasinac                 0.5   
3        45  154043.jpg             2.1    Maitland-Niles                 2.2   
4        48   39476.jpg             3.0  Papastathopoulos                 1.3   

   special squad_number status  team  team_code  total_points  transfers_in  \
0    False         None      i     1          3            43         24995   
1    False         None      a     1          3            44        205825   
2    False         None      a     1          3            54         64831   
3    False         None      a     1          3            40        610492   
4    False         None      a     1          3            57        182079   

   transfers_in_event  transfers_out  transfers_out_event value_form  \
0                 123          47166                 1600        0.3   
1                3438         158769                 3744        0.3   
2                1138         134155                  461        0.5   
3                1253         653265                  932        0.1   
4                 510         231215                  770        0.0   

  value_season        web_name  minutes  goals_scored  assists  clean_sheets  \
0          8.4         Mustafi     1205             0        2             4   
1          8.5        Bellerín     1156             1        0             4   
2         10.4       Kolasinac     1661             0        2             4   
3          8.9  Maitland-Niles     1292             0        2             3   
4         11.9        Sokratis     1696             2        0             4   

   goals_conceded  own_goals  penalties_saved  penalties_missed 

In [16]:
elements_df.columns

Index(['chance_of_playing_next_round', 'chance_of_playing_this_round', 'code',
       'cost_change_event', 'cost_change_event_fall', 'cost_change_start',
       'cost_change_start_fall', 'dreamteam_count', 'element_type', 'ep_next',
       'ep_this', 'event_points', 'first_name', 'form', 'id', 'in_dreamteam',
       'news', 'news_added', 'now_cost', 'photo', 'points_per_game',
       'second_name', 'selected_by_percent', 'special', 'squad_number',
       'status', 'team', 'team_code', 'total_points', 'transfers_in',
       'transfers_in_event', 'transfers_out', 'transfers_out_event',
       'value_form', 'value_season', 'web_name', 'minutes', 'goals_scored',
       'assists', 'clean_sheets', 'goals_conceded', 'own_goals',
       'penalties_saved', 'penalties_missed', 'yellow_cards', 'red_cards',
       'saves', 'bonus', 'bps', 'influence', 'creativity', 'threat',
       'ict_index', 'influence_rank', 'influence_rank_type', 'creativity_rank',
       'creativity_rank_type', 'threat_rank'

In [17]:
elements_df = elements_df[['id','first_name','second_name','dreamteam_count',\
                          'element_type','form','news','news_added','now_cost',\
                           'points_per_game','selected_by_percent','status','team',\
                           'total_points','value_form','value_season','minutes',\
                          'goals_scored','assists','clean_sheets','goals_conceded',\
                           'own_goals','penalties_saved','penalties_missed','yellow_cards',\
                           'red_cards','saves','bonus','bps']]

elements_df.set_index('id',inplace=True)
elements_df.head()

first_name       second_name  dreamteam_count  element_type form  \
id                                                                    
1    Shkodran           Mustafi                0             2  1.7   
2      Héctor          Bellerín                0             2  1.8   
3        Sead         Kolasinac                0             2  2.8   
4     Ainsley    Maitland-Niles                1             2  0.3   
5    Sokratis  Papastathopoulos                1             2  0.0   

                                      news                   news_added  \
id                                                                        
1   Hamstring injury - Unknown return date  2020-07-18T22:30:20.553834Z   
2                                           2019-12-09T20:00:21.228098Z   
3                                           2020-02-23T18:30:13.672943Z   
4                                           2019-09-22T18:00:10.824841Z   
5                                           2020-06-18T18:00:15.974146Z   

    now_cost points_per_game selected_by_percent status  team  total_points  \
id                                                                            
1         51             2.9                 0.4      i     1            43   
2         52             2.9                 1.0      a     1            44   
3         52             2.2                 0.5      a     1            54   
4         45             2.1                 2.2      a     1            40   
5         48             3.0                 1.3      a     1            57   

   value_form value_season  minutes  goals_scored  assists  clean_sheets  \
id                                                                         
1         0.3          8.4     1205             0        2             4   
2         0.3          8.5     1156             1        0             4   
3         0.5         10.4     1661             0        2             4   
4         0.1          8.9     1292             0        2             3   
5         0.0         11.9     1696             2        0             4   

    goals_conceded  own_goals  penalties_saved  penalties_missed  \
id                                                                 
1               17          0                0                 0   
2               18          0                0                 0   
3               25          0                0                 0   
4               20          0                0                 0   
5               25          0                0                 0   

    yellow_cards  red_cards  saves  bonus  bps  
id                                              
1              2          0      0      2  242  
2              2          0      0      4  204  
3              4          0      0      1  325  
4              4          1      0      3  229  
5              6          0      0      5  305

### Mapping teams and positions

In [18]:
elements_df['position'] = elements_df.element_type.map(element_types_df.singular_name_short)

In [19]:
elements_df[['second_name','position']]

second_name position
id                            
1             Mustafi      DEF
2            Bellerín      DEF
3           Kolasinac      DEF
4      Maitland-Niles      DEF
5    Papastathopoulos      DEF
..                ...      ...
587              Buur      DEF
607             Giles      DEF
609     Cavaco Jordao      MID
619   Castelo Podence      MID
652           Campana      FWD

[665 rows x 2 columns]

In [20]:
elements_df['team_name'] = elements_df.team.map(teams_df.name)

In [21]:
elements_df.drop('team',inplace=True,axis=1)

In [22]:
elements_df[['second_name','team_name']]

second_name team_name
id                             
1             Mustafi   Arsenal
2            Bellerín   Arsenal
3           Kolasinac   Arsenal
4      Maitland-Niles   Arsenal
5    Papastathopoulos   Arsenal
..                ...       ...
587              Buur    Wolves
607             Giles    Wolves
609     Cavaco Jordao    Wolves
619   Castelo Podence    Wolves
652           Campana    Wolves

[665 rows x 2 columns]

In [23]:
elements_df.drop('element_type',inplace=True,axis=1)

In [24]:
elements_df = elements_df[['first_name','second_name','position','team_name','dreamteam_count',\
                          'form','news','news_added','now_cost',\
                           'points_per_game','selected_by_percent','status',\
                           'total_points','value_form','value_season','minutes',\
                          'goals_scored','assists','clean_sheets','goals_conceded',\
                           'own_goals','penalties_saved','penalties_missed','yellow_cards',\
                           'red_cards','saves','bonus','bps']]
elements_df.head()

first_name       second_name position team_name  dreamteam_count form  \
id                                                                         
1    Shkodran           Mustafi      DEF   Arsenal                0  1.7   
2      Héctor          Bellerín      DEF   Arsenal                0  1.8   
3        Sead         Kolasinac      DEF   Arsenal                0  2.8   
4     Ainsley    Maitland-Niles      DEF   Arsenal                1  0.3   
5    Sokratis  Papastathopoulos      DEF   Arsenal                1  0.0   

                                      news                   news_added  \
id                                                                        
1   Hamstring injury - Unknown return date  2020-07-18T22:30:20.553834Z   
2                                           2019-12-09T20:00:21.228098Z   
3                                           2020-02-23T18:30:13.672943Z   
4                                           2019-09-22T18:00:10.824841Z   
5                                           2020-06-18T18:00:15.974146Z   

    now_cost points_per_game selected_by_percent status  total_points  \
id                                                                      
1         51             2.9                 0.4      i            43   
2         52             2.9                 1.0      a            44   
3         52             2.2                 0.5      a            54   
4         45             2.1                 2.2      a            40   
5         48             3.0                 1.3      a            57   

   value_form value_season  minutes  goals_scored  assists  clean_sheets  \
id                                                                         
1         0.3          8.4     1205             0        2             4   
2         0.3          8.5     1156             1        0             4   
3         0.5         10.4     1661             0        2             4   
4         0.1          8.9     1292             0        2             3   
5         0.0         11.9     1696             2        0             4   

    goals_conceded  own_goals  penalties_saved  penalties_missed  \
id                                                                 
1               17          0                0                 0   
2               18          0                0                 0   
3               25          0                0                 0   
4               20          0                0                 0   
5               25          0                0                 0   

    yellow_cards  red_cards  saves  bonus  bps  
id                                              
1              2          0      0      2  242  
2              2          0      0      4  204  
3              4          0      0      1  325  
4              4          1      0      3  229  
5              6          0      0      5  305

### Let us remove the players who played for 0 minutes this season

In [25]:
elements_df = elements_df[elements_df['minutes']!=0]
elements_df

first_name       second_name position team_name  dreamteam_count form  \
id                                                                          
1     Shkodran           Mustafi      DEF   Arsenal                0  1.7   
2       Héctor          Bellerín      DEF   Arsenal                0  1.8   
3         Sead         Kolasinac      DEF   Arsenal                0  2.8   
4      Ainsley    Maitland-Niles      DEF   Arsenal                1  0.3   
5     Sokratis  Papastathopoulos      DEF   Arsenal                1  0.0   
..         ...               ...      ...       ...              ...  ...   
420    Leander        Dendoncker      MID    Wolves                2  4.2   
473      Jesús    Vallejo Lázaro      DEF    Wolves                0  0.0   
487    Patrick           Cutrone      FWD    Wolves                0  0.0   
528      Pedro        Lomba Neto      MID    Wolves                0  1.3   
619     Daniel   Castelo Podence      MID    Wolves                1  3.2   

                                                  news  \
id                                                       
1               Hamstring injury - Unknown return date   
2                                                        
3                                                        
4                                                        
5                                                        
..                                                 ...   
420                                                      
473                            Loan deal ended on 24/1   
487  Joined Fiorentina on an 18-month loan deal on ...   
528                                                      
619                                                      

                      news_added  now_cost points_per_game  \
id                                                           
1    2020-07-18T22:30:20.553834Z        51             2.9   
2    2019-12-09T20:00:21.228098Z        52             2.9   
3    2020-02-23T18:30:13.672943Z        52             2.2   
4    2019-09-22T18:00:10.824841Z        45             2.1   
5    2020-06-18T18:00:15.974146Z        48             3.0   
..                           ...       ...             ...   
420  2019-10-11T10:00:13.063158Z        44             2.7   
473  2020-01-24T13:00:17.740017Z        46             0.5   
487  2020-01-10T20:00:21.473387Z        54             1.9   
528  2020-07-07T13:00:16.320894Z        49             2.0   
619                         None        49             3.0   

    selected_by_percent status  total_points value_form value_season  minutes  \
id                                                                              
1                   0.4      i            43        0.3          8.4     1205   
2                   1.0      a            44        0.3          8.5     1156   
3                   0.5      a            54        0.5         10.4     1661   
4                   2.2      a            40        0.1          8.9     1292   
5                   1.3      a            57        0.0         11.9     1696   
..                  ...    ...           ...        ...          ...      ...   
420                 8.5      a            99        1.0         22.5     2700   
473                 0.2      u             1        0.0          0.2      162   
487                 0.4      n            23        0.0          4.3      292   
528                 0.2      a            57        0.3         11.6      875   
619                 0.1      a            24        0.7          4.9      254   

     goals_scored  assists  clean_sheets  goals_conceded  own_goals  \
id                                                                    
1               0        2             4              17          0   
2               1        0             4              18          0   
3               0        2             4              25          0   
4               0        2     

## Exploratory Data Analysis

1) Form over fixture: All the players with the top form : form rises much quicker than it falls
2) ICT Index
3) Value_season
4) Points this week
5) Saves - goalkeepers
6) Assists - defenders
7) 

In [29]:
elements_df.shape

(511, 28)

Shape attribute of a dataframe returns the dimensionality, nothing but the number of rows and the number of columns of the dataframe. Therefore, there have been a total of 511 players across 20 teams that played in the English Premier League 2019-2020 season. 

In [31]:
elements_df.describe()

dreamteam_count    now_cost  total_points      minutes  goals_scored  \
count       511.000000  511.000000    511.000000   511.000000    511.000000   
mean          0.849315   52.228963     59.125245  1427.385519      1.894325   
std           1.474886   12.641839     47.697248  1018.103361      3.501902   
min           0.000000   38.000000      1.000000     2.000000      0.000000   
25%           0.000000   44.000000     19.000000   553.000000      0.000000   
50%           0.000000   49.000000     49.000000  1292.000000      1.000000   
75%           1.000000   55.000000     90.000000  2289.500000      2.000000   
max           9.000000  125.000000    232.000000  3330.000000     23.000000   

          assists  clean_sheets  goals_conceded   own_goals  penalties_saved  \
count  511.000000    511.000000      511.000000  511.000000       511.000000   
mean     1.708415      4.524462       21.547945    0.064579         0.033268   
std      2.530424      4.079467       15.611495    0.246023         0.209736   
min      0.000000      0.000000        0.000000    0.000000         0.000000   
25%      0.000000      1.000000        7.000000    0.000000         0.000000   
50%      1.000000      4.000000       20.000000    0.000000         0.000000   
75%      2.000000      7.000000       33.000000    0.000000         0.000000   
max     22.000000     17.000000       66.000000    1.000000         2.000000   

       penalties_missed  yellow_cards   red_cards       saves       bonus  \
count        511.000000    511.000000  511.000000  511.000000  511.000000   
mean           0.039139      2.440313    0.086106    4.266145    4.628180   
std            0.222364      2.608406    0.307461   19.293289    6.149662   
min            0.000000      0.000000    0.000000    0.000000    0.000000   
25%            0.000000      0.000000    0.000000    0.000000    0.000000   
50%            0.000000      2.000000    0.000000    0.000000    2.000000   
75%            0.000000      4.000000    0.000000    0.000000    7.000000   
max            2.000000     12.000000    2.000000  136.000000   38.000000   

              bps  
count  511.000000  
mean   262.213307  
std    204.992729  
min     -1.000000  
25%     85.500000  
50%    226.000000  
75%    400.500000  
max    918.000000

The describe method returns statistical quantities for all the numerical columns present in the dataframe. For example, lets look at the now_cost column. A player at an average costs 5.2 million and there exists a standard deviation of 1.2 million in the cost distribution, which signifies that the distribution is pretty narrow. In simpler words, most of the players in FPL have their values very close to the average cost 5.2 million. Looking at the 75% quantile, we can verify this fact. 75% of the players cost less than or equal to 5.5 million. In contrast, observe the total_points column. Having a high standard deviation of 47.69, the values are more loosely spread in the distribution.

Beware of using the describe method without any caution. This method would also include all those attributes in your dataframe which might only associate with certain rows. For example, the saves column would only associate with the goalkeepers in our dataset, but the statistical quantities given here would be calculated across all the players in the dataset: defenders, midfielders and forwards too. Well, this would not have been a problem if everyone had Kyle Walker abilities, would it?

Lets try to isolate the goalkeepers and then use the describe method

In [32]:
elements_df[elements_df.position == 'GKP'].describe()

dreamteam_count   now_cost  total_points      minutes  goals_scored  \
count         37.00000  37.000000     37.000000    37.000000          37.0   
mean           1.00000  47.081081     73.837838  1799.648649           0.0   
std            1.20185   5.775323     58.145657  1317.524915           0.0   
min            0.00000  39.000000      1.000000     7.000000           0.0   
25%            0.00000  43.000000     15.000000   435.000000           0.0   
50%            1.00000  46.000000     80.000000  1794.000000           0.0   
75%            2.00000  52.000000    126.000000  3150.000000           0.0   
max            4.00000  62.000000    169.000000  3330.000000           0.0   

         assists  clean_sheets  goals_conceded  own_goals  penalties_saved  \
count  37.000000     37.000000       37.000000  37.000000        37.000000   
mean    0.081081      5.513514       27.162162   0.108108         0.459459   
std     0.276725      5.080582       19.394320   0.314800         0.649555   
min     0.000000      0.000000        0.000000   0.000000         0.000000   
25%     0.000000      1.000000        7.000000   0.000000         0.000000   
50%     0.000000      5.000000       26.000000   0.000000         0.000000   
75%     0.000000      9.000000       39.000000   0.000000         1.000000   
max     1.000000     15.000000       63.000000   1.000000         2.000000   

       penalties_missed  yellow_cards  red_cards       saves      bonus  \
count              37.0     37.000000  37.000000   37.000000  37.000000   
mean                0.0      0.702703   0.054054   58.918919   5.891892   
std                 0.0      0.996239   0.229243   44.314143   6.181441   
min                 0.0      0.000000   0.000000    0.000000   0.000000   
25%                 0.0      0.000000   0.000000   15.000000   0.000000   
50%                 0.0      0.000000   0.000000   59.000000   4.000000   
75%                 0.0      1.000000   0.000000   96.000000  10.000000   
max                 0.0      4.000000   1.000000  136.000000  23.000000   

              bps  
count   37.000000  
mean   366.351351  
std    276.044100  
min      3.000000  
25%     86.000000  
50%    401.000000  
75%    633.000000  
max    777.000000

See the difference? saves now have a mean value of 58 per goalkeeper. Now obviously, that is across the whole season.

Now, if your aim is to become a Data Engineer and run all kinds of fancy Machine Learning models on your dataset, 'Null' or 'NA' is probably the scariest value for you. In order to get a holistic view of the null values present in your dataset, the isna() couples with any() methods does the job:

In [33]:
elements_df.isna().any()

first_name             False
second_name            False
position               False
team_name              False
dreamteam_count        False
form                   False
news                   False
news_added              True
now_cost               False
points_per_game        False
selected_by_percent    False
status                 False
total_points           False
value_form             False
value_season           False
minutes                False
goals_scored           False
assists                False
clean_sheets           False
goals_conceded         False
own_goals              False
penalties_saved        False
penalties_missed       False
yellow_cards           False
red_cards              False
saves                  False
bonus                  False
bps                    False
dtype: bool

Aha! We found the one culprit, the news_added column. The rest are all nice, having no missing data. Basically, isna() is a method which returns True/False value for each and every cell in your dataframe. Appending any() would then return a True value per column, if any of the rows in a particular column have atleast one null value. 

Lets see how many rows in the news_added column have null values:

In [34]:
elements_df.isna().sum()

first_name              0
second_name             0
position                0
team_name               0
dreamteam_count         0
form                    0
news                    0
news_added             83
now_cost                0
points_per_game         0
selected_by_percent     0
status                  0
total_points            0
value_form              0
value_season            0
minutes                 0
goals_scored            0
assists                 0
clean_sheets            0
goals_conceded          0
own_goals               0
penalties_saved         0
penalties_missed        0
yellow_cards            0
red_cards               0
saves                   0
bonus                   0
bps                     0
dtype: int64

83 rows out of the total 511 have null values in the news_added column.

In [46]:
len(elements_df.loc[528].news)

0

#### Most PPG?

In [28]:
elements_df.points_per_game.sort_values(ascending = False)

KeyError: "None of [Index(['8.2', '7.0', '6.8', '6.5', '6.3', '6.3', '6.1', '6.1', '5.8', '5.8',\n       ...\n       '0.9', '0.9', '0.8', '0.8', '0.8', '0.8', '0.8', '0.5', '0.5', '0.3'],\n      dtype='object', length=511)] are in the [columns]"

### FDR Test

In [117]:
from selenium import webdriver

In [118]:
d = webdriver.Chrome('/usr/local/bin/chromedriver')
d.get('https://fantasy.premierleague.com/transfers')


In [119]:
d.find_elements_by_xpath('//*[@id="root"]/div[2]/div/div/div/div[2]/div/table/tbody/tr[1]/td[1]/button').send_keys('\n')


In [120]:
d.find_element_by_xpath('//*[@id="root-dialog"]/div/dialog/div/div[2]/div[2]/div/ul/li[2]/a').send_keys('\n')

In [108]:
from bs4 import BeautifulSoup as soup
import re

In [126]:
name = soup(d.page_source, 'html.parser').find('h2', {'class': re.compile(r'ElementDialog__ElementHeading*')}).text
rating = soup(d.page_source, 'html.parser').find('td', {'class': re.compile(r'FixtureDifficulty*')}).text

In [128]:
print(name)

Nick Pope


In [203]:
d1 = webdriver.Chrome('/usr/local/bin/chromedriver')
d1.get('https://fantasy.premierleague.com/')

In [204]:
d1.find_element_by_xpath('//*[@id="loginUsername"]').send_keys('akshay.parkala@gmail.com')
d1.find_element_by_xpath('//*[@id="loginPassword"]').send_keys('Aksh@par07')
d1.find_element_by_xpath('//*[@id="root"]/div[2]/div/div/div[1]/form/div[3]/button').send_keys('\n')

In [205]:
d1.find_element_by_xpath('//*[@id="root"]/div[1]/div/div/div/nav/ul/li[4]/a').send_keys('\n')

In [206]:
info = d1.find_elements_by_xpath('//*[@id="root"]/div[2]/div/div/div/div[2]/div/table/tbody//td[1]/button')

In [ ]:
fdr = {}
key = True
while(True):
    for i in info:
        i.send_keys('\n')
        while(key):
            try:
                d1.find_element_by_xpath('//*[@id="root-dialog"]/div/dialog/div/div[2]/div[2]/div/ul/li[2]/a').send_keys('\n')
                key = False
            except:
                print("Fixture button Error")
                key = True
        team = soup(d1.page_source, 'html.parser').find('div', {'class': re.compile(r'ElementDialog__Club*')}).text
        print(team)
        try:
            rating = soup(d1.page_source, 'html.parser').find('td', {'class': re.compile(r'FixtureDifficulty__StyledFixtureDifficulty*')}).text
            print(rating)
            fdr[team] = rating
        except:
            print("Error")
            rating = 0
    if (len(fdr) == 20):
        break
    d1.find_element_by_xpath('//*[@id="root"]/div[2]/div/div/div/div[2]/div/div[2]/button[3]').send_keys('\n')
    info = d1.find_elements_by_xpath('//*[@id="root"]/div[2]/div/div/div/div[2]/div/table/tbody//td[1]/button')


In [208]:
fdr

{'Burnley': '2',
 'Wolves': '4',
 'Newcastle': '4',
 'Liverpool': '2',
 'Sheffield Utd': '2',
 'Man City': '2',
 'Man Utd': '3',
 'Chelsea': '4',
 'Spurs': '3',
 'Everton': '2',
 'Aston Villa': '2',
 'Leicester': '4',
 'Southampton': '3',
 'Arsenal': '2',
 'Watford': '4',
 'Brighton': '3',
 'Bournemouth': '4',
 'Norwich': '4',
 'Crystal Palace': '3',
 'West Ham': '2'}

In [ ]:
//*[@id="root-dialog"]/div/dialog/div/div[2]/div[2]/div/ul/li[2]/a

In [202]:
x = {'1':1,'2':2}
len(x)

2